File: merge_sort.go
package main

import "fmt"


func merge(arr []int, l, m, r int) {
	// find the length of left and right array
	n1 := m-l+1
	n2 := r-m

	// create temp arrays
	L := make([]int, n1)
	R := make([]int, n2)

	// Copy the elements over to temp array
	for i := 0; i < n1; i++ {
		L[i] = arr[l+i]
	}
	for j := 0; j < n2; j++ {
		R[j] = arr[m+1+j]
	}

	// Start merging the left and right array
	i, j := 0, 0
	k := l

	for i < n1 && j < n2 {
		if L[i] <= R[j] {
			arr[k] = L[i]
			i++
		} else {
			arr[k] = R[j]
			j++
		}
		k++
	}

	// copy remaining elements of L[]
	for i < n1 {
		arr[k] = L[i]
		i++
		k++
	}

	// copy remaining elements of R[]
	for j < n2 {
		arr[k] = R[j]
		j++
		k++
	}
}

func sort(arr []int, l, r int) {
	if l < r {
		mid := (l+r)/2

		sort(arr, l, mid)
		sort(arr, mid+1, r)

		merge(arr, l, mid, r)
	}
	fmt.Println(arr)
}


func main() {
	arr := []int{38,27,43,3,9,82,10}
	l := 0
	r := len(arr) - 1
	sort(arr, l, r)
}

File: python/algorithms/sorting/merge_sort.py
# Time complexity: O(n log n)
# Auxilary space: O(n) # because of recursion
from __future__ import annotations


def mergeSort(arr):
    if len(arr) > 1:

        # find mid
        mid = len(arr) // 2

        L = arr[:mid]  # left array

        R = arr[mid:]  # right array

        mergeSort(L)
        mergeSort(R)

        # once we splitted the array down to a single item. we can start comparing left and right array element and put it in a new array
        # declare 3 variables i,j,k
        # i is used to track L array items
        # j is used to track R array items
        # k is used to insert the sorted item in a new array
        i = j = k = 0

        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        # The below 2 loops are used to sort if there are any remaining items
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1


if __name__ == "__main__":
    arr = [38, 27, 43, 3, 9, 82, 10]
    mergeSort(arr)
    print(arr)
